server/src/main/java/org/opensearch/autotagging/Rule.java

```diff
@@ -0,0 +1,377 @@
+            validationException.addValidationError("Rule update time is not a valid epoch");
+        }
+        if (attributeMap != null && feature != null) {
+            validateAttributeMap(attributeMap, feature, validationException);
+        }
+    }
+
+    public static void requireNonNullOrEmpty(String value, String message, ValidationException validationException) {
+        if (value == null || value.isEmpty()) {
+            validationException.addValidationError(message);
+        }
+    }
+
+    public static void validateAttributeMap(
+        Map<Attribute, Set<String>> attributeMap,
+        Feature feature,
+        ValidationException validationException
+    ) {
+        for (Map.Entry<Attribute, Set<String>> entry : attributeMap.entrySet()) {
+            Attribute attribute = entry.getKey();
+            Set<String> attributeValues = entry.getValue();
+            if (!feature.isValidAttribute(attribute)) {
+                validationException.addValidationError(
+                    attribute.getName() + " is not a valid attribute name under the feature: " + feature.getName()
+                );
+            }
+            if (attributeValues.size() > MAX_NUMBER_OF_VALUES_PER_ATTRIBUTE) {
+                validationException.addValidationError(
+                    "Each attribute can only have a maximum of 10 values. The input attribute " + attribute + " exceeds this limit."
+                );
+            }
+            for (String attributeValue : attributeValues) {
+                if (attributeValue.isEmpty() || attributeValue.length() > MAX_CHARACTER_LENGTH_PER_ATTRIBUTE_VALUE_STRING) {
+                    validationException.addValidationError(
+                        "Attribute value [" + attributeValue + "] is invalid (empty or exceeds 100 characters)"
+                    );
+                }
+            }
+        }
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeString(description);
+        out.writeMap(attributeMap, Attribute::writeTo, StreamOutput::writeStringCollection);
+        out.writeString(label);
+        out.writeString(updatedAt);
+        out.writeString(feature.getName());
+    }
+
+    public static Rule fromXContent(final XContentParser parser) throws IOException {
+        return Builder.fromXContent(parser).build();
+    }
+
+    public String getDescription() {
+        return description;
+    }
+
+    public String getLabel() {
+        return label;
+    }
+
+    public String getUpdatedAt() {
+        return updatedAt;
+    }
+
+    public Feature getFeature() {
+        return feature;
+    }
+
+    public Map<Attribute, Set<String>> getAttributeMap() {
+        return attributeMap;
+    }
+
+    /**
+     * This enum enumerates the features that can use the Rule Based Auto-tagging
+     * @opensearch.experimental
+     */
+    public enum Feature {
```

Comment:
[NIT] Better to create a different file for this IMO. Same for below `Attribute`